SwiftGit2/Repository.swift

```diff
@@ -558,4 +579,246 @@ final public class Repository {
 	}
+
+	// MARK: - Diffs
+
+	public func diff(for commit: Commit) -> Result<Diff, NSError> {
+		typealias Delta = Diff.Delta
+
+		guard !commit.parents.isEmpty else {
+			// Initial commit in a repository
+			return self.diff(from: nil, to: commit.oid)
+		}
+
+		var diffs = [OpaquePointer]()
+		var mergeDiff: OpaquePointer? = nil
+		for parent in commit.parents {
+			let diff: Result<OpaquePointer, NSError> = self.diff(from: parent.oid, to: commit.oid)
+			guard diff.error == nil else {
+				return Result<Diff, NSError>.failure(diff.error!)
+			}
+
+			diffs.append(diff.value!)
+
+			if mergeDiff == nil {
+				mergeDiff = diff.value!
+			} else {
+				let mergeResult = git_diff_merge(mergeDiff, diff.value)
+				guard mergeResult == GIT_OK.rawValue else {
+					return .failure(NSError(gitError: mergeResult, pointOfFailure: "git_diff_merge"))
+				}
+			}
+		}
+
+		for diff in diffs {
+			git_object_free(diff)
+		}
+
+		return .success(Diff(mergeDiff!))
+	}
+
+	/// Caller is responsible to free returned git_diff with git_object_free
+	private func diff(from oldCommitOid: OID?, to newCommitOid: OID?) -> Result<OpaquePointer, NSError> {
+		assert(oldCommitOid != nil || newCommitOid != nil, "It is an error to pass nil for both the oldOid and newOid")
+
+		var oldTree: OpaquePointer? = nil
+		if let oid = oldCommitOid {
+			let result = unsafeTreeForCommitId(oid)
+			guard result.error == nil else {
+				return Result.failure(result.error!)
+			}
+
+			oldTree = result.value
+		}
+
+		var newTree: OpaquePointer? = nil
+		if let oid = newCommitOid {
+			let result = unsafeTreeForCommitId(oid)
+			guard result.error == nil else {
+				return Result.failure(result.error!)
+			}
+
+			newTree = result.value
+		}
+
+		var diff: OpaquePointer? = nil
+		let diffResult = git_diff_tree_to_tree(&diff,
+																					 self.pointer,
+																					 oldTree,
+																					 newTree,
+																					 nil)
+
+		git_object_free(oldTree)
+		git_object_free(newTree)
+
+		guard diffResult == GIT_OK.rawValue else {
+			return .failure(NSError(gitError: diffResult,
+															pointOfFailure: "git_diff_tree_to_tree"))
+		}
+
+		return Result<OpaquePointer, NSError>.success(diff!)
+	}
+
+	/// Memory safe
+	private func diff(from oldCommitOid: OID?, to newCommitOid: OID?) -> Result<Diff, NSError> {
+		assert(oldCommitOid != nil || newCommitOid != nil, "It is an error to pass nil for both the oldOid and newOid")
+
+		var oldTree: Tree? = nil
+		if oldCommitOid != nil {
+			let result = safeTreeForCommitId(oldCommitOid!)
+			guard result.error == nil else {
+				return Result<Diff, NSError>.failure(result.error!)
+			}
+			oldTree = result.value
+		}
+
+		var newTree: Tree? = nil
+		if newCommitOid != nil {
+			let result = self.safeTreeForCommitId(newCommitOid!)
+			guard result.error == nil else {
+				return Result<Diff, NSError>.failure(result.error!)
+			}
+			newTree = result.value!
+		}
+
+		if oldTree != nil && newTree != nil {
```

Comment:
You should be able to replace this with a `switch`.

```swift
switch (oldTree, newTree) {
case let (oldTree?, newTree?): // I'm not sure if you'll be able to reuse the names here
    // has both
case let (tree?, nil), let (nil, tree?):
   // has one
case (nil, nil):
    // has neither
}

Reply:
nice!

Reply:
It looks like `case let (oldTree?, newTree?)` works, the only problem I have is I'm not sure how to use tree here: `case let (tree?, nil), let (nil, tree?):`. I'm not sure how to use this. I need to know which of the two has a value.

Reply:
Ah, it looks like those will need to be separate cases. I didn't see that they were subtly different. Maybe we can reuse some code between them?